nomad/client_csi_endpoint.go

```diff
@@ -20,74 +21,101 @@ type ClientCSI struct {
 
 func (a *ClientCSI) ControllerAttachVolume(args *cstructs.ClientCSIControllerAttachVolumeRequest, reply *cstructs.ClientCSIControllerAttachVolumeResponse) error {
 	defer metrics.MeasureSince([]string{"nomad", "client_csi_controller", "attach_volume"}, time.Now())
-	// Get a Nomad client node for the controller
-	nodeID, err := a.nodeForController(args.PluginID, args.ControllerNodeID)
+
+	clientIDs, err := a.clientIDsForController(args.PluginID)
 	if err != nil {
-		return err
+		return fmt.Errorf("controller attach volume: %v", err)
 	}
-	args.ControllerNodeID = nodeID
 
-	// Get the connection to the client
-	state, ok := a.srv.getNodeConn(args.ControllerNodeID)
-	if !ok {
-		return findNodeConnAndForward(a.srv, args.ControllerNodeID, "ClientCSI.ControllerAttachVolume", args, reply)
-	}
+	for _, clientID := range clientIDs {
+		args.ControllerNodeID = clientID
+		state, ok := a.srv.getNodeConn(clientID)
+		if !ok {
+			return findNodeConnAndForward(a.srv,
+				clientID, "ClientCSI.ControllerAttachVolume", args, reply)
+		}
 
-	// Make the RPC
-	err = NodeRpc(state.Session, "CSI.ControllerAttachVolume", args, reply)
-	if err != nil {
+		err = NodeRpc(state.Session, "CSI.ControllerAttachVolume", args, reply)
+		if err == nil {
+			return nil
+		}
+		if a.isRetryable(err, clientID, args.PluginID) {
+			a.logger.Debug("failed to reach controller on client %q: %v", clientID, err)
+			continue
+		}
 		return fmt.Errorf("controller attach volume: %v", err)
 	}
-	return nil
+	return fmt.Errorf("controller attach volume: %v", err)
 }
 
 func (a *ClientCSI) ControllerValidateVolume(args *cstructs.ClientCSIControllerValidateVolumeRequest, reply *cstructs.ClientCSIControllerValidateVolumeResponse) error {
 	defer metrics.MeasureSince([]string{"nomad", "client_csi_controller", "validate_volume"}, time.Now())
 
-	// Get a Nomad client node for the controller
-	nodeID, err := a.nodeForController(args.PluginID, args.ControllerNodeID)
+	clientIDs, err := a.clientIDsForController(args.PluginID)
 	if err != nil {
-		return err
+		return fmt.Errorf("validate volume: %v", err)
 	}
-	args.ControllerNodeID = nodeID
 
-	// Get the connection to the client
-	state, ok := a.srv.getNodeConn(args.ControllerNodeID)
-	if !ok {
-		return findNodeConnAndForward(a.srv, args.ControllerNodeID, "ClientCSI.ControllerValidateVolume", args, reply)
-	}
+	for _, clientID := range clientIDs {
+		args.ControllerNodeID = clientID
+		state, ok := a.srv.getNodeConn(clientID)
+		if !ok {
+			return findNodeConnAndForward(a.srv,
+				clientID, "ClientCSI.ControllerValidateVolume", args, reply)
+		}
 
-	// Make the RPC
-	err = NodeRpc(state.Session, "CSI.ControllerValidateVolume", args, reply)
-	if err != nil {
+		err = NodeRpc(state.Session, "CSI.ControllerValidateVolume", args, reply)
```

Comment:
We're using RPC magic strings, so we could share the code for this whole loop. `args` and `reply` could share an interface. nbd either way

Reply:
I had that in an early attempt but I ended up with the shared function either taking `interface{}` or adding dummy methods to `client/structs/csi.go` to enforce an interface on the client request/response structs (using the `ToCSIRequest()` method as the interface would require type parameterization). I wasn't really wild about either approach.

Reply:
ah `ToCSIRequest` was what I had in mind, forgot about the return types there. Carry on!